src/broadcast/broadcast.rs

```diff
@@ -248,16 +281,83 @@ impl<N: NodeIdT> Broadcast<N> {
         let hash = *p.root_hash();
 
         // Save the proof for reconstructing the tree later.
-        self.echos.insert(sender_id.clone(), p);
+        self.echos.insert(sender_id.clone(), EchoContent::Full(p));
+
+        let mut step = Step::default();
+
+        // Upon receiving `N - 2f` `Echo`s with this root hash, send `CanDecode`
+        if !self.can_decode_sent && self.count_echos_full(&hash) >= self.coding.data_shard_count() {
```

Comment:
Do we need to store `can_decode_sent` by hash?
Because we might have to send more than one of them, e.g. if the proposer is faulty, it might send a value _v<sub>1</sub>_ with hash _h<sub>1</sub>_ to _N - f_ nodes _and_ value _v<sub>2</sub>_ with hash _h<sub>2</sub>_ to _f + 1_ nodes, with the overlapping nodes all malicious. So you may end up receiving _f + 1_ `Echo`s with _h<sub>2</sub>_, even though in the end _v<sub>1</sub>_ will get accepted. In that case, you should send `CanDecode` with both hashes eventually.

Reply:
Yeah you are right. Will fix this.